flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/source/IcebergSource.java

```diff
@@ -0,0 +1,289 @@
+    } catch (IOException e) {
+      throw new RuntimeException("Failed to close table loader", e);
+    }
+  }
+
+  @Override
+  public Boundedness getBoundedness() {
+    return scanContext.isStreaming() ? Boundedness.BOUNDED : Boundedness.CONTINUOUS_UNBOUNDED;
+  }
+
+  @Override
+  public SourceReader<T, IcebergSourceSplit> createReader(SourceReaderContext readerContext) {
+    ReaderMetricsContext readerMetrics =
+        new ReaderMetricsContext(readerContext.metricGroup());
+    return new IcebergSourceReader<>(readerFunction, readerContext, readerMetrics);
+  }
+
+  @Override
+  public SplitEnumerator<IcebergSourceSplit, IcebergEnumeratorState> createEnumerator(
+      SplitEnumeratorContext<IcebergSourceSplit> enumContext) {
+    return createEnumerator(enumContext, null);
+  }
+
+  @Override
+  public SplitEnumerator<IcebergSourceSplit, IcebergEnumeratorState> restoreEnumerator(
+      SplitEnumeratorContext<IcebergSourceSplit> enumContext, IcebergEnumeratorState enumState) {
+    return createEnumerator(enumContext, enumState);
+  }
+
+  @Override
+  public SimpleVersionedSerializer<IcebergSourceSplit> getSplitSerializer() {
+    return IcebergSourceSplitSerializer.INSTANCE;
+  }
+
+  @Override
+  public SimpleVersionedSerializer<IcebergEnumeratorState> getEnumeratorCheckpointSerializer() {
+    return IcebergEnumeratorStateSerializer.INSTANCE;
+  }
+
+  private SplitEnumerator<IcebergSourceSplit, IcebergEnumeratorState> createEnumerator(
+      SplitEnumeratorContext<IcebergSourceSplit> enumContext,
+      @Nullable IcebergEnumeratorState enumState) {
+    Table table = loadTable(tableLoader);
+    SplitAssigner assigner;
+    if (enumState == null) {
+      assigner = assignerFactory.createAssigner();
+    } else {
+      LOG.info("Iceberg source restored {} splits from state for table {}",
+          enumState.pendingSplits().size(), table.name());
+      assigner = assignerFactory.createAssigner(enumState.pendingSplits());
+    }
+
+    if (scanContext.isStreaming()) {
+      // Ideally, operatorId should be used as the threadPoolName as Flink guarantees its uniqueness within a job.
+      // SplitEnumeratorContext doesn't expose the OperatorCoordinator.Context, which would contain the OperatorID.
+      // Need to discuss with Flink community whether it is ok to expose a public API like the protected method
+      // "OperatorCoordinator.Context getCoordinatorContext()" from SourceCoordinatorContext implementation.
+      // For now, <table name>-<random UUID> is used as the unique thread pool name.
+      ContinuousSplitPlanner splitPlanner = new ContinuousSplitPlannerImpl(
+          table, scanContext, table.name() + "-" + UUID.randomUUID());
+      return new ContinuousIcebergEnumerator(enumContext, assigner, scanContext, splitPlanner, enumState);
+    } else {
+      return new StaticIcebergEnumerator(enumContext, assigner, table, scanContext, enumState);
+    }
+  }
+
+  public static <T> Builder<T> builder() {
```

Comment:
You could use `builderForRowData` here also.

Reply:
When I tried to implement the RowData factory method, I found it kind of need to duplicate 3 of the ScanContext configs for the RowDataReaderFunction.

```
  public static Builder<RowData> builderForRowData(Configuration readConfig, Table table, ScanContext context) {
   ReaderFunction<RowData> readerFunction = new RowDataReaderFunction(readConfig, table.schema(), context.project(),
        context.nameMapping(), context.caseSensitive(), table.io(), table.encryption());
    return new Builder<RowData>()
        .readerFunction(readerFunction);
  }
```

Reply:
Personally, I actually like to make `ScanContext` public and exposed to users. Then it is ok to have method like `Builder<RowData> builderForRowData(Configuration readConfig, Table table, ScanContext context)`. We can also avoid duplicate more than a dozen of methods from `ScanContext `to `IcebergSource$Builder`. Users can construct ScanContext once.

Reply:
We can probably do better here. Why not initialize `readerFunction` in the `build` method if it is `null`?